stateMachine.go

```diff
@@ -0,0 +1,137 @@
+/*
+HSM provides the framework for Hierarchical State Machine implementations.
+
+Related Documents:
+    - State Driven Development: https://confluence.enflick.com/x/3YwuAg
+    - Go State Machine Framework: https://confluence.enflick.com/x/9AIgAw
+
+Included in this framework are the following components:
+
+  - StateMachine:
+    Machine that controls the event processing
+
+  - State:
+    Interface that must be implemented by all States in the StateMachine
+
+  - Transition:
+    Interface that is implemented by each of the different types of transitions:
+
+      - ExternalTransition:
+        Transition from current state to a different state.  On execution the following takes place:
+          1. OnExit is called on the current state and all parent states up to the parent state that owns
+             the new state (or the parent state is nil)
+          2. action() associated with the the transition is called
+          3. OnEnter() is called on the new state which may call OnEnter() for a sub-state.  The final
+             new current state is returned by the OnEnter() call
+
+      - InternalTransition:
+        Transition within the current state.  On execution the following takes place:
+          1. action() associated with the the transition is called
+
+      - EndTransition:
+        Transition from current state that terminates the state machine.  On execution the following takes place:
+          1. OnExit is called on the current state and all parent states until there are no more parent states
+          2. action() associated with the the transition is called
+
+  - Event:
+    An event represents something that has happened (login, logout, newCall, networkChange, etc.) that might drive
+    a change in the state machine
+*/
+package hsm
+
+import (
+	"context"
+	"go.uber.org/zap"
+)
+
+// StateMachine manages event processing as implemented by each State
+type StateMachine struct {
+	currentState State
+	svc          Service
+}
+
+// StateMachine constructor
+func NewStateMachine(svc Service, startState State) *StateMachine {
```

Comment:
I would not depend on the service having a logger, and pass the logger in as an explicit dependency here. It is cleaner when setting up and logging events in the state machine.

Reply:
Avoiding passing in a Service handle was actually a key design decision I made with the original state machine framework, as it forces a layer of knowledge about the usage of the framework into the framework itself. Looking through this I don't think we actually need to pass the Service in at all; any registered actions (which are today taking in the service) can be implemented as methods on the Service (which will prevent us needing to supply the Service to the action invocation). The Service is only used by Execute to supply to the Action so I think this change will clean this up.

Reply:
Okay.  So instead we should pass everything we need to our State constructor to get the most out of the compiler to find bugs?  For example, Logger, ESLConnection, RedisConnection, HTTPConnection, etc?

I think this makes sense.

Reply:
Yes exactly. Each state may take a different set of parameters; so we can be sure we're properly initializing each state. It also makes testing the states easier as they aren't dependent on a properly-set-up service. Since the state machine doesn't create any states itself, there could be possible state machine implementations where the service pattern is superfluous, or cases where multiple services may be desired, so keeping it up to the state to properly initialize its children will ensure the least amount of coupling between our state machine and the states that it manages.